module-apps/apps-common/widgets/DateSetSpinner.cpp

```diff
@@ -0,0 +1,267 @@
+    {
+        day->set_value(static_cast<unsigned>(date.day()));
+        month->set_value(static_cast<unsigned>(date.month()));
+        year->set_value(static_cast<int>(date.year()));
+        applySizeRestrictions();
+    }
+
+    void DateSetSpinner::applySizeRestrictions()
+    {
+        day->setMinimumWidthToFitText();
+        firstSlash->setMinimumWidthToFitText();
+        month->setMinimumWidthToFitText();
+        secondSlash->setMinimumWidthToFitText();
+        year->setMinimumWidthToFitText();
+
+        setMinimumSize(getWidgetMinimumAreaWidth(), getFontHeight(noFocusFontName));
+        setMaximumWidth(widgetMaximumArea.w);
+
+        HBox::informContentChanged();
+    }
+
+    std::uint32_t DateSetSpinner::getWidgetMinimumAreaWidth()
+    {
+        return day->widgetMinimumArea.w + firstSlash->widgetMinimumArea.w + month->widgetMinimumArea.w +
+               secondSlash->widgetMinimumArea.w + year->widgetMinimumArea.w;
+    }
+
+    template <typename Spinner> void DateSetSpinner::attachDateField(Spinner *&field, typename Spinner::range &&range)
+    {
+        field = new Spinner(std::move(range), Boundaries::Continuous);
+        updateFont(field, noFocusFontName);
+        field->setAlignment(Alignment(Alignment::Horizontal::Right, Alignment::Vertical::Center));
+        field->setEdges(RectangleEdge::None);
+        field->setPenFocusWidth(date_set_spinner::focus::size);
+        field->set_value(0);
+        addWidget(field);
+    }
+
+    void DateSetSpinner::attachSlash(gui::Label *&slash)
+    {
+        slash = new gui::Label(this, 0, 0, 0, 0);
+        updateFont(slash, noFocusFontName);
+        slash->setEdges(gui::RectangleEdge::None);
+        slash->setAlignment(gui::Alignment(gui::Alignment::Horizontal::Center, gui::Alignment::Vertical::Center));
+        slash->setEdges(RectangleEdge::None);
+        slash->setText("/");
+        slash->activeItem = false;
+    }
+
+    void DateSetSpinner::setFont()
+    {
+        updateFont(day, noFocusFontName);
+        updateFont(month, noFocusFontName);
+        updateFont(year, noFocusFontName);
+
+        setMinimumSize(getWidgetMinimumAreaWidth(), getFontHeight(noFocusFontName));
+
+        resizeItems();
+    }
+
+    std::uint16_t DateSetSpinner::getFontHeight(const std::string &fontName) const
+    {
+        const RawFont *font = FontManager::getInstance().getFont(fontName);
+        return font->info.line_height;
+    }
+
+    void DateSetSpinner::updateFont(TextFixedSize *elem, const std::string &fontName)
+    {
+        elem->setFont(fontName);
+        elem->setMinimumHeightToFitText();
+        elem->setMinimumWidthToFitText();
+
+        elem->setText(elem->getText());
+    }
+
+    bool DateSetSpinner::onInput(const InputEvent &inputEvent)
+    {
+        if (editMode != EditMode::Edit) {
+            return false;
+        }
+
+        if (auto ret = this->focusItem->onInput(inputEvent)) {
+            setMaxDay();
```

Comment:
are you sure we have to set the last day of the month?
Please rename the function to `setLastDay` if so.

Reply:
Yes Im sure, ok will rename to setLastDay

Reply:
currently as the automatic date correction works different the setLasDay is called in handleEnterKey()